### PR TITLE
[Gardening] add reference link for "iff"

### PR DESCRIPTION
**Pull request by takecian**
#### What's in this pull request?

expand "iff" to "if and only if" for readability improvment
This PR is update of #1467 
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by practicalswift** [comment]
@takecian There are a few more instances if you want to cover them all.

Run `git grep -i "[^a-z]iff[^a-z]"` to find all instances of "iff" :-)

**Comment by takecian** [comment]
@practicalswift Thanks for feedback!
I expand all "iff".

File: lib/SILGen/Condition.cpp [hunk 1]
```diff
@@ -39,7 +39,7 @@ static SILLocation getContinuationLoc(SILBasicBlock &BB, SILLocation Fallback) {
 void Condition::exitTrue(SILGenFunction &SGF, ArrayRef<SILValue> Args) {
   // If there's no continuation block, it's because the condition was
   // folded to true.  In that case, we just continue emitting code as
-  // if we were still in the true case, and we're unreachable iff the
+  // if we were still in the true case, and we're unreachable if and only if the
```
Comment (jtbandes) [inline lib/SILGen/Condition.cpp]: Should we care about re-wrapping comments like this?

Comment (OktayGardener) [inline lib/SILGen/Condition.cpp]: Why is this necessary?
If you're a programmer at Apple or a contributer to one of the biggest open source languages, hopefully you've taken any math or computer science class and can understand what _iff_ means.

#gardenerout

**Comment by lattner** [comment]
While I can understand the claim that "iff" should be known by everyone everywhere, this is empirically not actually true.  Not everyone is a native english speaker, and many who are are not familiar with the notion of iff in any case.  +1 from me for expanding these out.

**Comment by tkremenek** [comment]
Indeed.  I suggested this change in response to a different pull request where this confusion over iff was evident.

**Comment by takecian** [comment]
I added reference link for "iff" for document files from suggestion of #1467.
I keep it as is for "iff" in source files.
What do you think about it?

**Comment by OktayGardener** [comment]
Thanks guys, you've given me another view to look at it! The first time I saw _iff_ in my life, I thought it was a typo as well, so expanding was for the better! :dancer:

**Comment by dabrahams** [comment]
Are these really the only instances of iff in the stdlib source?

**Comment by takecian** [comment]
Yes, i also add reference for iff in the stdlib sources.

**Comment by takecian** [comment]
Hi, if any feedback, please let me know.

File: docs/proposals/ValueSemantics.rst [hunk 2]
```diff
@@ -287,10 +287,12 @@ distinction.  In fact both of the following must be present in order
...
 
 1. Use of (one of) the big three operations on an object ``x``,
-   creating shared mutable state iff ``x`` is a reference
+   creating shared mutable state [iff] ``x`` is a reference
 
 2. In-place mutation of ``x`` *while a (reference) copy is extant* and
-   thus can be observed through the copy iff ``x`` is a reference.
+   thus can be observed through the copy [iff] ``x`` is a reference.
+
+[iff]: https://en.wikipedia.org/wiki/If_and_only_if "If And Only If"
```
Comment (dabrahams) [inline docs/proposals/ValueSemantics.rst]: I don't think markdown link syntax is going to work in a .rst file.  Please rebuild the docs to confirm that you at least have valid syntax.

File: docs/proposals/rejected/Clonable.rst [hunk 3]
```diff
@@ -87,7 +87,9 @@ Implications for Equality
 
 We think the programming model suggested for generics has some pretty
 strong implications for equality of ``class``\ es: ``a == b`` must
-return true iff ``a`` and ``b`` refer to the same object.
+return true [iff] ``a`` and ``b`` refer to the same object.
+
+[iff]: https://en.wikipedia.org/wiki/If_and_only_if "If And Only If"
```
Comment (dabrahams) [inline docs/proposals/rejected/Clonable.rst]: Markdown link in .rst file

File: docs/proposals/valref.rst [hunk 4]
```diff
@@ -284,7 +284,9 @@ type parameter, as follows::
 
    ref x : T // always declares a ref
    val x : T // always declares a val
-   var x : T // declares a val iff T is a val
+   var x : T // declares a val [iff] T is a val
+
+[iff]: https://en.wikipedia.org/wiki/If_and_only_if "If And Only If"
```
Comment (dabrahams) [inline docs/proposals/valref.rst]: Markdown link in .rst file

**Comment by dabrahams** [comment]
I don't see any changes to swift files in this PR, which, unlike #1467, only touches 5 files.  Did you forget to push your changes?

**Comment by takecian** [comment]
Oh, I will check and get back to you.

**Comment by takecian** [comment]
I've pushed commits which,
- Expand "iff" to "if and only if" in rst files.
- Add reference link for "iff" in source files.

**Comment by dabrahams** [comment]
@swift-ci Please test

**Comment by dabrahams** [comment]
@swift-ci Please test

**Comment by takecian** [comment]
Sorry, I missed pushing one commit which resolves conflict.
Now I've pushed.

**Comment by takecian** [comment]
@swift-ci Please test

File: CHANGELOG.md [hunk 5]
```diff
@@ -2512,9 +2512,11 @@ Swift 2.2
     AnyObject>`.  It bridges in O(1), without memory allocation.
 
   - `Dictionary<K, V>` has an implicit conversion to `NSDictionary`.
-    `Dictionary<K, V>` bridges to `NSDictionary` iff both `K` and `V` are
+    `Dictionary<K, V>` bridges to `NSDictionary` [iff] both `K` and `V` are
```
Comment (jrose-apple) [inline CHANGELOG.md]: This is not proper Markdown; I think you mean `[iff][]`.

Comment (takecian) [inline CHANGELOG.md]: Thanks for feedback.
I think that this also works. Please see,
https://github.com/takecian/swift/blob/master/CHANGELOG.md

Comment (jrose-apple) [inline CHANGELOG.md]: Hm, it's not valid [original-Markdown](http://daringfireball.net/projects/markdown/syntax#link), but it is in CommonMark (and clearly in GitHub-Flavored Markdown). I'd still prefer using the "collapsed" form with the empty `[]`, though, please!

Comment (dabrahams) [inline CHANGELOG.md]: > On Mar 9, 2016, at 5:26 PM, Jordan Rose <notifications@github.com> wrote:
> 
> I'd still prefer using the "collapsed" form with the empty [], though, please!
> 
Why is that better?

Comment (jrose-apple) [inline CHANGELOG.md]: It makes it clearer (to me) that it's a link rather than some kind of placeholder, or an array that ought to be in backticks. But maybe I've just used Markdown long enough to have habits and preferences, which doesn't automatically make them good ones.

**Comment by takecian** [comment]
I added empty brackets in markdown file due to suggestion from @jrose-apple .

**Comment by jrose-apple** [comment]
Ah, that comment applied to *all* uses of that syntax, not just CHANGELOG.md. Either we care or we don't.

**Comment by jrose-apple** [comment]
(I do see that Doxygen explicitly [calls it out as legal](http://www.stack.nl/~dimitri/doxygen/manual/markdown.html#md_reflinks).

**Comment by takecian** [comment]
I applied to all [iff] with empty brackets.

**Comment by rintaro** [comment]
@jrose-apple
At least for Xcode, `[iff]` (without `[]`) is preferred in this doc: https://developer.apple.com/library/prerelease/mac/documentation/Xcode/Reference/xcode_markup_formatting_ref/LinkReference.html#//apple_ref/doc/uid/TP40016497-CH56-SW1

**Comment by jrose-apple** [comment]
So it is! Thanks for finding that, @rintaro, and, um, sorry for the trouble, @takecian. Feel free to erase those last two commits via force push.

**Comment by dabrahams** [comment]
@takecian please resolve conflicts and remove all the empty brackets, with our apologies for Jordan’s misplaced zeal ;-)

**Comment by takecian** [comment]
No problem! I'll resolve conflicts.

**Comment by takecian** [comment]
I found a lot of conflicts on master when I started update.
So I restart changing document on another branch. I will open another PR.

**Comment by takecian** [comment]
I created new PR for this. Please move to #1633.
Thanks.
